src/rmt.rs

```diff
@@ -0,0 +1,691 @@
+                    carrier_level: carrier.carrier_level as u32,
+                    carrier_duty_percent: carrier.duty_percent.0,
+                    idle_output_en: config.idle.is_some(),
+                    idle_level: config.idle.map(|i| i as u32).unwrap_or(0),
+                    // Bug? This doesn't seem to work for longer length data on ESP32S2.
+                    loop_en,
+                    // Bug? When looping is working, it seems to be ignored at least on my ESP32S2.
+                    loop_count,
+                },
+            },
+        };
+
+        unsafe {
+            esp!(rmt_config(&sys_config))?;
+            esp!(rmt_driver_install(C::channel(), 0, 0))?;
+        }
+
+        Ok(Self { pin, channel })
+    }
+
+    /// Get speed of the channel’s internal counter clock.
+    ///
+    /// This calls [rmt_get_counter_clock()][rmt_get_counter_clock]
+    /// internally. It is used for calculating the number of ticks per second for pulses.
+    ///
+    /// See [Pulse::new_with_duration()].
+    ///
+    /// [rmt_get_counter_clock]: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html#_CPPv421rmt_get_counter_clock13rmt_channel_tP8uint32_t
+    pub fn counter_clock(&self) -> Result<Hertz, EspError> {
+        let mut ticks_hz: u32 = 0;
+        esp!(unsafe { rmt_get_counter_clock(C::channel(), &mut ticks_hz) })?;
+        Ok(ticks_hz.into())
+    }
+
+    /// Start sending the given signal without blocking.
+    ///
+    /// `signal` is captured for safety so that the user can't change the data while transmitting.
+    pub fn start<S>(&self, signal: S) -> Result<(), EspError>
+    where
+        S: Signal,
+    {
+        self.write_items(&signal, false)
+    }
+
+    /// Start sending the given signal while blocking.
+    pub fn start_blocking<S>(&self, signal: &S) -> Result<(), EspError>
+    where
+        S: Signal,
+    {
+        self.write_items(signal, true)
+    }
+
+    fn write_items<S>(&self, signal: &S, block: bool) -> Result<(), EspError>
+    where
+        S: Signal,
+    {
+        let items = signal.as_slice();
+        esp!(unsafe { rmt_write_items(C::channel(), items.as_ptr(), items.len() as i32, block,) })
+    }
+
+    /// Stop transmitting.
+    pub fn stop(&self) -> Result<(), EspError> {
+        esp!(unsafe { rmt_tx_stop(C::channel()) })
+    }
+
+    /// Stop transmitting and release the driver.
+    ///
+    /// This will return the pin and channel.
+    pub fn release(self) -> Result<(P, C), EspError> {
+        self.stop()?;
+        esp!(unsafe { rmt_driver_uninstall(C::channel()) })?;
+        Ok((self.pin, self.channel))
+    }
+
+    pub fn set_looping(&self, looping: config::Loop) -> Result<(), EspError> {
+        esp!(unsafe { rmt_set_tx_loop_count(C::channel(), looping.into()) })
+    }
+}
+
+/// Signal storage for [`Transmit`] in a format ready for the RMT driver.
+pub trait Signal {
```

Comment:
I'm still not sure what is the best solution for modeling signals... we are kind of leaking a private `esp-idf-sys` type (`rmt_item32_t`) into the signature of this method. But with that said, I currently don't have any better ideas, so let's leave it like that, and if we come up with better modeling options in future, we can always adjust the code. :)

Reply:
I did consider to use something like `[bitvec](https://docs.rs/bitvec/latest/bitvec/index.html)::[array](https://docs.rs/bitvec/latest/bitvec/array/index.html#)` but that would add an unnecessary dependency.